crates/walrus-sui/src/client/retry_client/failover.rs

```diff
@@ -148,29 +159,48 @@ impl<ClientT, BuilderT: LazyClientBuilder<ClientT> + std::fmt::Debug>
                 "only one client available, try adding rpc urls".to_string(),
             ));
         }
-
         let mut state = self.state.lock().await;
+        self.fetch_next_client_inner(tried_client_indices, &mut state)
+            .await
+    }
 
+    /// Fetches the next client set in the FailoverWrapper. Excluding the clients that have already
+    /// been tried in the `tried_client_indices` set.
+    ///
+    /// This function is locked on the `state` mutex, and will update the state if a new client is
+    /// fetched.
+    async fn fetch_next_client_inner(
+        &self,
+        tried_client_indices: &mut BTreeSet<usize>,
+        state: &mut Option<FailoverState<ClientT>>,
+    ) -> Result<Arc<ClientT>, FailoverError> {
         // Check if the state of the FailoverWrapper has already advanced to a client we haven't
         // tried yet. If so, go ahead and use that one. Note that this condition is subtle as it
         // mutates the `tried_client_indices`.
-        if tried_client_indices.len() < self.max_tries
-            && tried_client_indices.insert(state.current_index)
-        {
-            return Ok(state.client.clone());
-        }
+        if let Some(state) = state.as_mut() {
```

Comment:
```suggestion
        if let Some(state) = state.as_ref() {
```
No need for `as_mut` here as far as I can tell?

Reply:
Could also coalesce this `if let` with the next statement (assignment of `next_index`).

Reply:
Done. Good catch!